source/opt/folding_rules.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_FOLDING_RULES_H_
+#define LIBSPIRV_UTIL_FOLDING_RULES_H_
+
+#include "constants.h"
+#include "def_use_manager.h"
+#include "ir_builder.h"
+#include "ir_context.h"
+
+#include <cstdint>
+#include <vector>
+
+namespace spvtools {
+namespace opt {
+
+using FoldingRule =
+    std::function<bool(ir::Instruction*, const std::vector<uint32_t>&,
+                       const std::vector<const analysis::Constant*>&)>;
+
+class FoldingRules {
+ public:
+  FoldingRules();
+
+  const std::vector<FoldingRule>& GetRulesForOpcode(SpvOp opcode) {
+    static std::vector<FoldingRule> empty_vector;
```

Comment:
I find it a bit dubious to use a static variable here.  What's the benefit over making this a regular data member that is initialized during construction?

Actually, this goes against the style guide.  static vars should be POD.

Reply:
Made it a class member.